linode/manager
# fix: [M3-6001] - Fix notification menu displaying empty menu on secondary status click

## Description 📝

**What does this PR do?**
Currently, when a linode has a secondary status (e.g. snapshot, migrating), clicking on the status opens up an empty notification menu. This happens on the entity detail page as well as the landing page. This PR fixes the above behavior so clicking on a status opens the notification menu with visible menu contents. 

###  Bug Cause  🐛 
 Relying on two different booleans for the menu opening, one within `notificationContext` and the other based on whether the Menu `isExpanded`. Since the NotificationButton and NotificationMenu did not know about context, the menu was not actually considered `open` on secondary status click to render its contents. 

###  Bug Fix ✅ 
 Rather than rely on `isExpanded`, the visibility of the NotificationMenu is now toggled based on the value of `notificationContext.menuOpen`.  

**Notes:**
- This PR also refactors the notification menu from the ReachUI Menu components to MUI v5 Menu components. This allowed us to remove some workaround code like the `handleAnyClick` function, which was causing difficulties when trying to implement this fix with the Reach UI menu.
- I removed the Storybook story for NotificationMenu, as it was somehow broken and not showing up in Storybook anyway. If the team prefers, I can create a follow up ticket for adding a story for the updated NotificationMenu component.

TODO:
- [x] Fix menu failing to scroll when there are many events and menu not staying at the top of the screen -- this is a positioning/overflow issue and fixing one has resulted in causing the other
- [x] Fix failing unit test somehow caused by changes to `Button.tsx` and E2E failure on `smoke-linode-landing-table.spec.ts`.

## Preview 📷

🐛  **Prod:**

https://user-images.githubusercontent.com/114685994/226739521-8a4635e7-0cae-4d61-b25c-40146b773b06.mov

✅  **Fix:**

https://user-images.githubusercontent.com/114685994/226739544-0c742610-0c25-4c62-b3bd-340b4d0840d6.mov

## How to test 🧪

**What are the steps to reproduce the issue or verify the changes?**

**Test menu opening/closing:**
1. Kick off a process with a secondary status like migrating or taking a backup of a linode. Powering off/on a linode also works, but you have to be quick.
2. Wait for that process to start and then click on the secondary status of that linode from the linode landing page. 
3. Observe that the notification menu opens and displays contents. 
4. View the linode details page of that same linode and click on the secondary status.
5. Observe that the notification menu opens and displays contents. 
6. Click the notification button (bell icon) to open the notification menu and click the icon again to close it. Confirm that behavior has not changed from prod.
7. Use the keyboard to navigate to the bell icon and test opening the drawer via both `Enter` and `Space` (use `Esc` to close). Confirm that behavior has not changed from prod.
8. Test all the various ways to open and close the notification menu in succession and make sure that menu responds as expected.
9.  Check that the notification icon badge is visible/hidden as expected when events are viewed.

**Test the notification events view:**
1. Notifications but a few (<5) events (turn on mocks)
2. No notifications and no events (comment out code inside of `serverHandlers.ts`)

**Check for UI consistency:**
1. Test smaller screen sizes to make sure there are no regressions from prod.
2. Check that notification menu button styling and notification menu styling look the same as prod.
3. Check that when the menu is open and the page scrolls, the menu stays in place at the top of the screen (no `fixed` position).
4. Check that when number of notifications/events exceeds the max menu height (turn on MSW and expand the notifications), the user can scroll in the menu. (TODO: fix)

**How do I run relevant unit or e2e tests?**
```
yarn cy:run -s "cypress/e2e/notificationsAndEvents/notifications.spec.ts"
```
Broke when modifying `Button.tsx`: 
```
yarn test LinodePowerControl.test.tsx
```
